docs/_tutorials/automatic-tensor-parallelism.md

```diff
@@ -0,0 +1,129 @@
+
+```python
+# create the model
+import transformers
+import deepspeed
+pipe = pipeline("text2text-generation", model="google/t5-v1_1-small", device=local_rank)
+# Initialize the DeepSpeed-Inference engine
+pipe.model = deepspeed.init_inference(
+    pipe.model,
+    mp_size=world_size,
+    dtype=torch.float
+)
+output = pipe('Input String')
+```
+
+Previously, to run inference with only tensor parallelism for the models that we don't support kernels, you can pass an injection policy that shows the two specific linear layers on a Transformer Encoder/Decoder layer: 1) the attention output GeMM and 2) layer output GeMM. We need these parts of the layer to add the required all-reduce communication between GPUs to merge the partial results across model-parallel ranks. Below, we bring an example that shows how you can use deepspeed-inference with a T5 model:
+
+```python
+# create the model
+import transformers
+from transformers.models.t5.modeling_t5 import T5Block
+import deepspeed
+pipe = pipeline("text2text-generation", model="google/t5-v1_1-small", device=local_rank)
+# Initialize the DeepSpeed-Inference engine
+pipe.model = deepspeed.init_inference(
+    pipe.model,
+    mp_size=world_size,
+    dtype=torch.float,
+    injection_policy={T5Block: ('SelfAttention.o', 'EncDecAttention.o', 'DenseReluDense.wo')}
+)
+output = pipe('Input String')
+```
+
+With automatic tensor parallelism, we do not need to provide the injection policy for supported models. The injection policy will be determined at runtime and applied automatically.
+
+
+## Example Script
+
+We can observe performance improvement with automatic tensor parallism using the [inference test suite](https://github.com/microsoft/DeepSpeedExamples/blob/master/inference/huggingface/text-generation/inference-test.py). The script includes per token latency, bandwidth, throughput and memory checks for comparison.
+
+
+## Launching
+
+To run without DeepSpeed and without tensor parallelism:
+
+```bash
+deepspeed --num_gpus <num_gpus> DeepSpeedExamples/inference/huggingface/text-generation/inference-test.py --name <model> --batch_size <batch_size>
+```
+
+
+To enable tensor parallelism, you need to set the `ds_inference` to True for the compatible models:
+
+```bash
+deepspeed --num_gpus <num_gpus> DeepSpeedExamples/inference/huggingface/text-generation/inference-test.py --name <model> --batch_size <batch_size> --ds_inference
+```
+
+## OPT 13B Inference Performance Comparison
+
+The following results were collected using V100 SXM2 32GB GPUs.
```

Comment:
(from discussion) let's add the flag(s) used to reproduce the numbers below. Should we also add mem bandwidth below to the table?

Reply:
Hi @jeffra, I included the test_performance flag. I didn't include the bandwidth numbers because I felt Tflops captured enough information for the comparison. I also didn't save the bandwidth numbers but I could calculate them from the Tflops results.